mock-resources-as-code/chart.yml

```diff
@@ -0,0 +1,7 @@
+apiVersion: v1beta
+kind: chart
+name: beautiful-chart
+spec:
+  query: funny-query
+
+  type: bar
```

Comment:
> Pros of a full json spec for a visualisation:
> - open-source, completely portable

@owlas I agree with this I feel like keeping the version controlled code/config as close to the source as possible is ideal. One challenge I see, and forgive me if this is incorrect since I have no frontend experience, but I was looking at the [echarts getting started guide](https://echarts.apache.org/handbook/en/get-started/) and it looks like it starts to include data in the config. This is the example I'm seeing, are these the options that feed the json spec that youre referring to?

```var option = {
        title: {
          text: 'ECharts Getting Started Example'
        },
        tooltip: {},
        legend: {
          data: ['sales']
        },
        xAxis: {
          data: ['Shirts', 'Cardigans', 'Chiffons', 'Pants', 'Heels', 'Socks']
        },
        yAxis: {},
```

In this case the x axis includes all distinct clothing categories but if the data changes and "jeans" gets added then my next export of the chart will have a diff even without me making any edits to it. Or a dev warehouse with a subset of the data would have a different config than prod. Am I correct in my understanding? Any thoughts on how to solve that case? If so, maybe it can be abstracted somehow?

I was also experimenting with the export chart as json button that exists in the Lightdash UI already and I saw some similar data based diffs, is that using the full json config that youre thinking about using here?

>Cons:
> - complex, even simple charts will have big specs

I actually dont think theres a problem with having large specs as long as its easy to read the diff. It might be obvious but making sure theres no unordered lists or anything like that where exporting an unchanged chart would result in diffs. I've run into this when I tried to use other BI tools that have export functionality and the experience ended up resembling more of a backup because basically all the resources get clobbered whether theyre changed or not, it serves the purpose of backing up to git but it makes it impossible to understand what changed and merging two branches wouldnt be possible.

Reply:
Hey @pnadolny13 !

Yes the export we use embeds the data but when we generate the echarts config we use the `datasets` api to separate the data from the chart encoding.

There's an example of the dataset api here, which is useful: https://echarts.apache.org/examples/en/editor.html?c=dataset-encode0

This way, the chart configuration only needs to contain references to field names and don't need to embed the actual values.

Reply:
@owlas Awesome - Looking forward to testing this out!